### ISSUE TITLE
Incorrect character side image flashes into view at (seemingly) random places

### ISSUE THREAD

**Issue by aemixx** [renpy/renpy#5131]
I'm still having what looks like the same issue as was described here https://github.com/renpy/renpy/issues/4828, despite using a version of Ren'py that includes [the associated fix ](https://github.com/renpy/renpy/commit/a1bcd63ba92827371bfaa2ef4904d2df6c7829d6)(8.1.3)! 

See the bug in the video below--the player character side image (different from our default narrator, which has no image) flashes into view between a different characters's lines. This happens very rarely, and won't always happen under seemingly the same set of circumstances.

Compared to the previous issue linked above, we have side sprites fade in rather than slide, but otherwise it seems to be the same issue. In my case, it didn't start happening after saving at a menu, and it went away after loading a save that was created while the bug was visible. 

I also am not able to reliably reproduce the bug, and saving at a choice menu doesn't cause it to happen (perhaps thanks to [the fix here](https://github.com/renpy/renpy/commit/a1bcd63ba92827371bfaa2ef4904d2df6c7829d6)). The same player who recorded this video used Skip to get to the same point in the game, and the bug didn't reappear.

https://github.com/renpy/renpy/assets/27410008/68fdbf7b-8d6f-4294-a9d9-9ed4fa06ec50

Our characters are linked to side images this way:
```
define narrator = Character(name=None, ctc = "ctc_blink")
define althea = Character("Althea", what_prefix="\"", what_suffix="\"", who_color="#FFFFFF", who_outlines=[ (3, "#db6a0740", 0, 0), (2, "#db6a0780", 0, 0),  (1, "#db6a07bf", 0, 0) ], image="althea", ctc = "ctc_blink")
```
The side image is defined this way, referencing the full character sprite:
```
image side althea = LayeredImageProxy("althea", Transform (crop=(72, 0, 450, 420)))
```
The side image is defined in the say screen this way:
```
screen say(who, what):
    style_prefix "say"

    window:
        id "window"

        if who is not None:

            window:
                id "namebox"
                style "namebox"
                text who id "who"

        text what id "what"


    ## If there's a side image, display it above the text. 
    add SideImage() xalign 0.0 yalign 1.0
```
Transitions between side images are defined this way:
```
# sprite transitions
transform same_transform(old, new):
    old
    new with Dissolve(0.2, alpha=True)

define config.side_image_same_transform = same_transform
define config.say_attribute_transition = dissolve 
define config.side_image_only_not_showing = True #side image only if sprite not on screen
define config.side_image_null = Null(width=540, height=350)

transform change_transform(old, new):
    contains: # this version has the sprite dissolve in and out
        old
        alpha 1.0
        yalign 1.0
        xpos 0.0 xanchor 0.0
        linear 0.2 alpha 0.0

    contains:
        new
        alpha 0.0
        yalign 1.0
        xpos 0.0 xanchor 0.0
        linear 0.2 alpha 1.0
```

**Comment by aemixx** [comment]
Update: In the save file where the bug appeared, even though it seemed to go away after exiting and loading the save, the bug reappeared at the next choice menu. However, in a save where the bug had never appeared, getting to that same choice menu did not cause the bug to appear.

**Comment by aemixx** [comment]
Update: in a save file with a bug, at a point when the bug appears, if the `scene`, `show`, `camera:` and `with fade` lines there are all commented out, the bug no longer appears at this point. Then, if only the `scene` line is uncommented, the bug once again appears at this point.

This is the point in the code where the bug appears, with a side image flashing into view briefly between the two lines of narration:

```
    "I must start delivering the bread I brought to those who can eat it."
    $ quick_menu = False
    scene bg medievalmedbay givebread:
        truecenter() zpos -1500 zzoom True zoom 1.0
    show farkas medievalmedbay givebread:
        truecenter() zpos -750 zzoom True zoom 1.0
    show judith medievalmedbay givebread:
        truecenter() zpos -450 zzoom True zoom 1.0
    camera:
        xpos -0.15625 ypos -0.13888888888 zpos -450
    with fade
    $ quick_menu = True
    "I do so, without question. This was obviously what I came to do."
```

**Comment by aemixx** [comment]
Update: In a save file with the bug, the incorrect side image only flashes into view between some lines of dialog/narration and not others. For example:

```
farkas_narration trunks glow_trunks eyes_closed mouth_thinking "I wake up from a dreamless, restless sleep."
# no bug here
farkas_narration eyes_thinking mouth_parted "Someone's poking my side...? Oh, it's Hekla."
# no bug here
show hekla closeup mouth_talking:
    xanchor 0.5 xpos 0.45 alpha 0
    easein 0.5 xpos 0.5 alpha 1
hekla "Hey... There's someone downstairs who wants to speak to you."
# BUG HERE!
show hekla eyes_raisebrow with dissolve
hekla "He says he's Hektor's father."
```

So it only appears between the lines with a `show` statement that includes a transition at the end such as `with dissolve.` If I edit this line to remove the `with dissolve` at the end, the bug no longer appears then.

This is different from my previous comment above, when removing the `with fade` line at the end did not stop the bug from appearing so long as the `scene` statement was still uncommented.

**Comment by aemixx** [comment]
Update: I tried setting the narrator side image to use a fully transparent `narrator.png`, like so, but the bug still appeared at the same times in the affected save file afterward:
`define narrator = Character(name=None, image="narrator", ctc = "ctc_blink")`

Adding `if who is not None` as a conditional within our say screen like this also seems to have no affect on the bug appearing at all the same places as usual within an affected save file.
```
if who is not None:
        add SideImage() xalign 0.0 yalign 1.0
```

However, completely commenting out the `add SideImage()` line *does* get rid of the bug (and all side images in the entire game), but that at least tells me that the issue is finding the correct conditional statement to put after the `if` and who=None isn't necessarily true when the bug occurs.

If we find a conditional statement that works, I would still call that a workaround rather than a true fix of the root of the bug, but maybe it would point us in the right direction to find the true root.

**Comment by aemixx** [comment]
After more experimentation, it seems like `who` isn't `None` when the bug appears in between dialog or narration lines. `who` is actually equal to the name of the character who speaks *after* the glitch appears, and the glitch appears during the transition.

In a save file affected by the glitch, the glitch seems to happen whenever any displayable onscreen goes through a transition that involves `dissolve`, `fade`, or any other transition that isn't instantaneous, regardless of what side images are used for the lines of dialog or narration before and after the glitch.

**Comment by aemixx** [comment]
To test, I defined a new transition that takes zero time:
```
define dissolvefast = Dissolve(0.0)
```
But when I use this in place of dissolve, the glitch still appears. It flashes into view very briefly. If I increase the time of the transition:
```
define dissolvefast = Dissolve(1.0)
```
Then the glitched side image remains visible for a full second before we see the correct side image and the text or the next line.

So something is happening during dissolve and fade, even if their times are set to zero, that allows the incorrect side sprite to appear for a moment.

But not if it's a Dissolve() happening within `same_transform` or `change_transform` for the side image, only a dissolve or fade for a regular displayable on screen.

**Comment by aemixx** [comment]
I may have found a workaround! The underlying engine bug is still there, but I can for now sidestep it this way:

Even though `if who is not None` before `add SideImage()` didn't completely prevent side sprites from appearing during transitions and scene statements as described above, `if who is not "Narrator"` worked after I gave the narrator character a name. I also set the narrator's image to a transparent png for good measure:

`define narrator = Character(name="Narrator", image="narrator", ctc = "ctc_blink")`

Note that I then had to make sure the narrator also wouldn't have a nameplate appear, now that it had a name. So the full say screen becomes:
```
screen say(who, what):
style_prefix "say"

window:
    id "window"
    if who is not None and who is not "Narrator":
        window:
            id "namebox"
            style "namebox"
            text who id "who"
    text what id "what"
if who is not "Narrator":
    add SideImage() xalign 0.0 yalign 1.0
```

But otherwise that seemed to work!

I wonder if what makes my game different and caused an issue is that we have "narrator" versions of our player characters which have `name=None` but still have side images, and then somehow via the name being None, the side image got associated with the narrator as well?

That's a guess because like I said, whatever originally caused the issue is likely still a bug in the engine--otherwise, I wouldn't have needed to change from the default say screen. I hope this helps to find it!

**Comment by the66F95** [comment]
if you see the effect only after loading a save file, try setting `_side_image_attributes` in the `after_load` label to `None`.
```py
label after_load:
    $ _side_image_attributes = None
    return
```
it won't fix what's causing the issue but you should get rid of the extra side image.

**Comment by aemixx** [comment]
It never appeared only after loading in my experience. Sometimes loading would make it go away and then it would come back at the next menu screen.

**Comment by solarProtag** [comment]
Hello I'm continuing this because I've been running into the same issue and managed to replicate it in a blank-slate renpy game to narrow down the issue. 

https://github.com/renpy/renpy/assets/36377994/ac9fc391-5bf8-4951-a514-97e6bd5f1b22

![image](https://github.com/renpy/renpy/assets/36377994/4132ba8a-3db8-48d4-b1f9-073021136df3)

```
define h = Character("hanayo")
image hanayo = "hanayo casual norm.png"

define t = Character("test", image="test")
layeredimage side test:

    attribute base default

    group outfit auto:

        attribute 1 default
        attribute 2

    group hair auto:

        attribute a default
        attribute b

    group face auto:

        attribute neutral default
        attribute happy

label start:

    scene black
    with fade

    t 1 neutral "test"

    scene bg room
    with fade
    
    t 2 happy "shirtB happy"

    show hanayo with dissolve

    t 1 b neutral "another test"

    return
```

Everything I posted here is me replicating this bug step by step, what turns up when I open the console and print what the current attributes are after loading, and the entire script of this empty game to show that there's nothing else at play here.

Hopefully this gets fixed soon!

**Comment by mal** [comment]
Confirmed that test case above no longer presents in `8.1.4.23112903+nightly` due to fix from [the66F95](https://github.com/the66F95).
